R/add_cols.R

```diff
@@ -0,0 +1,293 @@
+#' overwritten.
+#' @keywords internal
+#'
+NULL
+
+#' @name .add_date
+.add_date_contact <- function(.data,
+                              contact_type = c("first", "last"),
+                              distribution = c("pois", "geom"),
+                              ...,
+                              outbreak_start_date = NULL) {
+  contact_type <- match.arg(contact_type)
+  distribution <- match.arg(distribution)
+
+  stopifnot(
+    "outbreak_start_date is only required for adding date of last contact" =
+      contact_type == "last" && !is.null(outbreak_start_date) ||
+      contact_type == "first" && is.null(outbreak_start_date)
+  )
+
+  rdist <- switch(distribution,
+    pois = stats::rpois,
+    geom = stats::rgeom
+  )
+
+  # c() over ...length() to ensure NULL is not counted by length
+  if (length(c(...)) == 0) {
+    stop("Distribution parameters are missing, check config", call. = FALSE)
+  }
+
+  # name list elements with vec names to ensure arg matching in do.call
+  args <- c(nrow(.data), list(...))
+
+  contact_delay <- tryCatch(
+    do.call(rdist, args = args),
+    error = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    },
+    warning = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    }
+  )
+
+  if (contact_type == "first") {
+    .data$date_first_contact <- .data$date_last_contact - contact_delay
+  } else {
+    .data$date_last_contact <- .data$infector_time + contact_delay +
+      outbreak_start_date
+  }
+
+  # return data
+  .data
+}
+
+#' @name .add_date
+.add_hospitalisation <- function(.data,
+                                 onset_to_hosp,
+                                 hosp_risk) {
+  infected_lgl_idx <- .data$infected == "infected"
+  num_infected <- sum(infected_lgl_idx)
+  .data$hospitalisation <- NA_real_
+  .data$hospitalisation[infected_lgl_idx] <- .data$time[infected_lgl_idx] +
+    onset_to_hosp(num_infected)
+
+  # hosp_risk is either numeric or <data.frame> or NA
+  if (!rlang::is_lgl_na(hosp_risk)) {
+    if (is.numeric(hosp_risk)) {
+      # size is converted to an integer internally in sample()
+      pop_sample <- sample(
+        which(infected_lgl_idx),
+        replace = FALSE,
+        size = (1 - hosp_risk) * num_infected
+      )
+      .data$hospitalisation[pop_sample] <- NA_real_
+    } else {
+      for (i in seq_len(nrow(hosp_risk))) {
+        age_bracket <- hosp_risk$min_age[i]:hosp_risk$max_age[i]
+        age_group <- which(.data$age %in% age_bracket)
+        not_hosp_prob <- 1 - hosp_risk$risk[i]
+        # size is converted to an integer internally in sample()
+        age_group_sample <- sample(
+          age_group,
+          replace = FALSE,
+          size = not_hosp_prob * length(age_group)
+        )
+        .data$hospitalisation[age_group_sample] <- NA_real_
+      }
```

Comment:
Thanks for point this out. I think it can be vectorised but I'll need to give it some thought. I'll convert this into an issue and tackle it after the release.

Reply:
Logged in issue #125.